extension/lib/shopgate/CartExtensionPipeline.js

```diff
@@ -98,21 +123,51 @@ class ShopgateCartExtensionPipeline {
    * @return {Promise<boolean>}
    */
   async updateProducts (cartItems) {
+    const cartId = await this.getCartId()
+
+    // try to create a cart level lock as some other calls will depend on our result
+    const lock = await ShopgateConcurrency.create(this._context.storage.extension)
+      .lock(getCartLockName(cartId), 10000)
+
     let updateSuccess = true
-    await this._bigCommerceCartRepository.updateItems(
-      cartItems.map((item) => {
-        return BigCommerceCartRepository.createLineItemUpdate(item.CartItemId, item.quantity)
-      }),
-      (failureEvent) => {
-        this._context.log.error({
-          msg: 'Failed updating product',
-          reason: failureEvent.reason,
-          cartItemId: failureEvent.item.itemId,
-          quantity: failureEvent.item.quantity
-        })
-        updateSuccess = false
+    try {
+      await this._bigCommerceCartRepository.updateItems(
+        cartItems.map((item) => {
+          return BigCommerceCartRepository.createLineItemUpdate(item.CartItemId, item.quantity)
+        }),
+        (failureEvent) => {
+          this._context.log.error(decorateDebug({
+            reason: failureEvent.reason,
+            cartItemId: failureEvent.item.itemId,
+            quantity: failureEvent.item.quantity
+          }), 'Failed updating product')
+
+          this._messagesRepository.push(cartId, failureEvent.reason, failureEvent.item.itemId)
```

Comment:
I'm not sure, does this need to be awaited?

Reply:
let's talk

Reply:
I guess I'd say let's await.

Reply:
done